### PR TITLE
Consolidate output parameters in global.params

### PR DESCRIPTION
**Pull request by thewilsonator**
It provides a uniform interface for outputs with regard to the name of the file/directory/buffer/enabled status etc. and it reduces the number of variables in Params by 14 (+7/-21).

### CODE REVIEW

File: src/dmd/globals.d [hunk 1]
```diff
@@ -98,6 +90,25 @@ enum FeatureState : byte
     enabled = 1    /// Specified as `-preview=`
 }
 
+struct Output
+{
+    /// Configuration for the C++ header generator
+    enum Mode : uint
+    {
+        none,     /// Don't generate output
+        doOutput, /// Generate output
```
Comment (Geod24) [inline src/dmd/globals.d:99]:
```suggestion
        normal, /// Generate output
```

Would be more in line with `full`.

Comment (thewilsonator) [reply src/dmd/globals.d]: done

**Comment by RazvanN7** [comment]
What are the benefits of this change? Besides saving 9 LOC. I agree that it looks more compact, but having all those fields with individual names is much clearer IMHO.

**Comment by thewilsonator** [comment]
It provides a uniform interface for outputs with regard to the name of the file/directory/buffer/enabled status etc. and it reduces the number of variables in Params by 14 (+7/-21)

>  but having all those fields with individual names is much clearer IMHO.

they are not named consistently and they are clearly grouped.

**Comment by thewilsonator** [comment]
Yay, finally it's looking green.

File: src/dmd/globals.d [hunk 2]
```diff
@@ -205,27 +210,14 @@ extern (C++) struct Param
...
-    const(char)[] hdrname;               // write 'header' file to docname
-    bool hdrStripPlainFunctions = true; // strip the bodies of plain (non-template) functions
-
-    CxxHeaderMode doCxxHdrGeneration;      /// Generate 'Cxx header' file
-    const(char)[] cxxhdrdir;            // write 'header' file to docdir directory
-    const(char)[] cxxhdrname;           // write 'header' file to docname
-
-    bool doJsonGeneration;              // write JSON file
-    const(char)[] jsonfilename;          // write JSON file to jsonfilename
+    Output doc;                         // Generate embedded documentation comments
```
Comment (MoonlightSentinel) [inline src/dmd/globals.d:213]:
```suggestion
    Output ddoc;                         // Process embedded documentation comments
```

File: src/dmd/globals.d [hunk 3]
```diff
@@ -205,27 +210,14 @@ extern (C++) struct Param
...
+    Output hdr;                         // Generate `.di` 'header' files
```
Comment (MoonlightSentinel) [inline src/dmd/globals.d:214]:
```suggestion
    Output dHdr;                         // Generate `.di` 'header' files
```
(Or maybe `diHeader`).

For a clear seperation between the `.di` and `.h` generation

File: src/dmd/globals.d [hunk 4]
```diff
@@ -98,6 +90,19 @@ enum FeatureState : byte
...
+    bool doOutput;      // Output is enabled
+    bool fullOutput;    // Generate comments for hidden declarations (for -HC),
+                        // and don't strip the bodies of plain (non-template) functions (for -H)
+
+    const(char)[] dir;  // write to directory 'dir'
+    const(char)[] name; // write to file 'name'
+    Array!(const(char)*) files; // Other files associated with this output,
+                                // e.g. macro include files for Ddoc, dependencies for makedeps
+    OutBuffer* buffer;  // if this output is buffered, this is the buffer
+    int bufferLines;    // number of lines written to the buffer
```
Comment (MoonlightSentinel) [inline src/dmd/globals.d:104]: These members are not relevant for e.g. the C++ header generator.
Maybe `Output` should be split into `Configuration` (CLI arguments) and `State` (buffer, files, ...)

File: src/dmd/globals.h [hunk 5]
```diff
@@ -87,6 +79,26 @@ enum class FeatureState : signed char
...
+    /// Configuration for the compiler generator
+    enum class Mode : signed char
+    {
+        none,     /// Don't generate output
+        normal,   /// Generate output
+        full      /// Generate comments for hidden declarations (for -HC),
+                  /// and don't strip the bodies of plain (non-template) functions (for -H)
+    };
+    Mode mode;
+    bool doOutput() const;
```
Comment (MoonlightSentinel) [inline src/dmd/globals.h:93]: This declaration doesn't match the D code

**Review by MoonlightSentinel — COMMENTED** [review]
Otherwise LGTM

File: src/dmd/globals.d (hunk 4 above)
Comment (thewilsonator) [reply src/dmd/globals.d:104]: Maybe in a followup PR.

**Review by WalterBright — DISMISSED** [review]
> _no description provided._

C'mon, guys. A description is needed for refactoring:

1. what
2. why

**Comment by RazvanN7** [comment]
@WalterBright I have updated the description. @thewilsonator feel free to add anything else.

**Comment by RazvanN7** [comment]
@thewilsonator Are you willing to go through the trouble of rebasing this?

**Comment by thewilsonator** [comment]
> rebasing this?

Currently `frontend.h` is completely wrong but can be (re)generated. There's a couple of other issues that will need to be fixed post rebase.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**